x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java

```diff
@@ -202,33 +224,43 @@ protected void readInternal(final byte[] buffer, final int offset, final int len
                     if (e instanceof AlreadyClosedException || (e.getCause() != null && e.getCause() instanceof AlreadyClosedException)) {
                         try {
                             // cache file was evicted during the range fetching, read bytes directly from source
-                            bytesRead += readDirectly(pos, pos + len, buffer, off);
+                            bytesRead = readDirectly(pos, pos + len, buffer, off);
                             continue;
                         } catch (Exception inner) {
                             e.addSuppressed(inner);
                         }
                     }
                     throw new IOException("Fail to read data from cache", e);
 
+                } finally {
+                    totalBytesRead += bytesRead;
                 }
             }
-            assert bytesRead == length : "partial read operation, read [" + bytesRead + "] bytes of [" + length + "]";
+            assert totalBytesRead == length : "partial read operation, read [" + totalBytesRead + "] bytes of [" + length + "]";
+            stats.incrementBytesRead(lastReadPosition, position, totalBytesRead);
+            lastReadPosition = position + totalBytesRead;
         }
 
         int readCacheFile(FileChannel fc, long end, long position, byte[] buffer, int offset, long length) throws IOException {
             assert assertFileChannelOpen(fc);
-            return Channels.readFromFileChannel(fc, position, buffer, offset, Math.toIntExact(Math.min(length, end - position)));
+            int bytesRead = Channels.readFromFileChannel(fc, position, buffer, offset, Math.toIntExact(Math.min(length, end - position)));
+            stats.addCachedBytesRead(bytesRead);
+            return bytesRead;
         }
 
         @SuppressForbidden(reason = "Use positional writes on purpose")
         void writeCacheFile(FileChannel fc, long start, long end) throws IOException {
             assert assertFileChannelOpen(fc);
+            final String fileName = cacheFileReference.getFileName();
             final byte[] copyBuffer = new byte[Math.toIntExact(Math.min(COPY_BUFFER_SIZE, end - start))];
-            try (IndexInput input = in.openInput(cacheFileReference.getFileName(), ioContext)) {
+            logger.trace(() -> new ParameterizedMessage("writing range [{}-{}] of file [{}] to cache file", start, end, fileName));
```

Comment:
Ah, I think this will not include the shard ID so we might struggle to interpret the logs when there are lots of shards to search. I think logging `cacheFileReference` itself gives us everything we need.

Reply:
Makes sense, I pushed 9de84533884a13e397406ed16aa5638cccb8ea9a